pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -660,10 +660,8 @@ func (cm *containerManagerImpl) GetPluginRegistrationHandler() cache.PluginHandl
 // TODO: move the GetResources logic to PodContainerManager.
 func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error) {
 	opts := &kubecontainer.RunContainerOptions{}
-	// Set container annotations from the CDI reconciler to
-	// trigger CDI injection
 	if cm.draManager != nil {
-		annotations, err := cm.draManager.GetCDIAnnotations(pod, container)
+		annotations, err := cm.draManager.PrepareResources(pod, container)
```

Comment:
Will need to look more closely at this and what the implications are of having such a heavyweight call here.

Reply:
It's not a heavyweight call according to the KEP. Here is a quote from there:
```
NodePrepareResource
...
This operation SHALL do as little work as possible as it’s called after a pod is scheduled to a node. All potentially failing operations SHALL be done during allocation phase.
```

Reply:
It's heavy-weight in the sense that an API call over a gRPC interface needs to be made.

Reply:
Agree. We can stress test this if needed. So far it didn't break any tests and enabled retrying of the PrepareResource call. I'll be happy to move it to other place if needed.

Reply:
@klueska
> Will need to look more closely at this and what the implications are of having such a heavyweight call here.

After looking at the Kubelet code again I'd say we're safe with this call in this place. There are a lot more heavyweight things happen in the same area. Here is what I've learned:
Containers are created by the runtime manager in the [startContainer](https://github.com/kubernetes/kubernetes/blob/v1.25.0/pkg/kubelet/kuberuntime/kuberuntime_container.go#L173:37) method. It does everything needed for container creation:
- pulls the image
- generates container config (this is where we call PrepareDynamicResources)
- calls PreCreateContainer hook (CPU and Memory manager update containerConfig.Linux.Resources.CpusetCpus  and containerConfig.Linux.Resources.CpusetMems  in this hook)
- creates container using container config generated before by calling CRI API CreateContainer

Reply:
Another idea is to call PrepareResources also on admission like it was implemented before. In most cases it will be called there, but if it errors for some reason we can retry it on the container configuration phase.

Reply:
btw, DeviceManager does something similar. It calls Allocate method not only on admission, but also on container configuration phase: https://github.com/kubernetes/kubernetes/blob/v1.25.0/pkg/kubelet/cm/devicemanager/manager.go#L880

Reply:
Yes it does this, but if this ever actually gets called here, all topology alignment will be incorrect.

Reply:
That's true. If we want to support topology we'll have to call PrepareResources on admission stage. That would make it hard to repeat this call if it fails.
One way to do support both things (Topology and retrying) is to make admission retries. But that could be quite big change.

Reply:
Resolved. We won't implement topology support in DRA, at least for its alpha.